### PR TITLE
Cleanup of DOCTEST_DO_BINARY_EXPRESSION_COMPARISON. Fixes #651

### LINKED ISSUE #651: Comparison with implicit cast from non-const value can't be decomposed
## Description

If a class can be implicitly constructed from non-const ref and has comparison operators, the decomposer can't handle the decomposition when the the non-const ref is used.

### Steps to reproduce

```c++
struct int_pointer {
    int* p = nullptr;
    int_pointer() = default;
    
    // non-const ref used in constructor
    // we don't want to accidentally construct a dangling pointer from a temporary
    int_pointer(int& i) : p(&i) {} 
    
    explicit operator bool() { return !!p; }
    int val() const { return *p; }
};

struct int_holder {
    int i = 0;
    bool operator==(int_pointer p) const {
        if (!p) return false;
        return i == p.val();
    }
};

TEST_CASE("hp") {
    int_holder h{8};
    int i = 8;
    
    // this comparison is safe
    // int_pointer can be implicitly constructed from this int here
    CHECK(h == i);
    // CHECK((h == i)); - this compiles just fine with extra parens
}
```

### Extra information

This issue was introduced by in https://github.com/doctest/doctest/commit/02af672f0d3093157daecb55ce4e109ffaf2bc53

The change is here (pasting link as code, otherwise github squashes it): `https://github.com/doctest/doctest/commit/02af672f0d3093157daecb55ce4e109ffaf2bc53#diff-60b065fd4481eca8e17eb467e91b3f552ab8a7fb56d58abfbaf870a98cd432c6R1154` 

I have absolutely no idea why it was made. `const R&&` seems pretty suspicious. It seems to me that it can be just removed and the code can be reverted to how it was in 3.8.6:

```c++
#define DOCTEST_DO_BINARY_EXPRESSION_COMPARISON(op, op_str, op_macro)                              \
    template <typename R>                                                                          \
    DOCTEST_NOINLINE SFINAE_OP(Result,op) operator op(R&& rhs) {                             \
    bool res = op_macro(doctest::detail::forward<const L>(lhs), doctest::detail::forward<R>(rhs));                                                             \
        if(m_at & assertType::is_false)                                                            \
            res = !res;                                                                            \
        if(!res || doctest::getContextOptions()->success)                                          \
            return Result(res, stringifyBinaryExpr(lhs, op_str, rhs));                             \
        return Result(res);                                                                        \
    }                                                                                              \
```

### PR DESCRIPTION
**Pull request by iboB**
## Description

Removes suspicious change introduced in https://github.com/doctest/doctest/commit/02af672f0d3093157daecb55ce4e109ffaf2bc53

I have no idea why it was made and it seems pointless to me. 

## GitHub Issues

This fixes #651

### CODE REVIEW

**Comment by onqtam** [comment]
can confirm that this does not break the compilation for which issue #495 was originally reported for: https://godbolt.org/z/cx6xs4fT6

File: examples/all_features/decomposition.cpp [hunk 1]
```diff
@@ -23,3 +23,36 @@ TEST_CASE("Move Only Type") {
...
+
+DOCTEST_MSVC_SUPPRESS_WARNING_WITH_PUSH(4866) // eval order
+TEST_CASE("Impl cast from non-const value") {
+    int_holder h{ 8 };
+    int i = 8;
+
+    // this comparison is safe
+    // int_pointer can be implicitly constructed from this int here
+    CHECK(h == i);
+    CHECK((h == i));
```
Comment (onqtam) [inline examples/all_features/decomposition.cpp:56]: IMHO the extra parens should be removed

File: examples/all_features/decomposition.cpp [hunk 2]
```diff
@@ -23,3 +23,36 @@ TEST_CASE("Move Only Type") {
...
+    int i = 0;
+    bool operator==(int_pointer p) const {
+        if (!p) return false;
+        return i == p.val();
+    }
+};
+
+DOCTEST_MSVC_SUPPRESS_WARNING_WITH_PUSH(4866) // eval order
+TEST_CASE("Impl cast from non-const value") {
+    int_holder h{ 8 };
```
Comment (onqtam) [inline examples/all_features/decomposition.cpp:50]: either this
```c++
    int_holder h;
    h.i = 8;
```
or a ctor for int_holder

**Comment by Saalvage** [comment]
I knew I'd seen something like this before.. const r-values references have previously plagued doctest as well!
https://github.com/doctest/doctest/commit/ce13bc44b99c87e918b98bdadb17008d2dd54d55

**Comment by Saalvage** [comment]
@onqtam #495 does NOT work with this. Your godbolt link compiles using clang, while the original issue is present in gcc. But it seems like it's broken on dev as well.. Works fine on master though!

**Pull request merged**